File: legacy/src/logic/solvers/EulerCromerSolver.java
package logic.solvers;

import model.physics.Body2D;
import model.physics.Universe2D;

public 
class EulerCromerSolver
extends Physics2DSolver
{
	public 
	EulerCromerSolver(Universe2D u) 
	{
		super(u);
		timestep = 0.1;
		displayName = "Semi-implicit Euler (Euler-Cromer)";
	}
	
	@Override
	public Physics2DSolver getNewInstance(Universe2D u) {
		return new EulerCromerSolver(u);
	}

	@Override public 
	void step() 
	{
		computeAllBodiesAccelerations();
		
		//velocity & position loop
		for(Body2D b1 : universe.bodies)
		{
			b1.velocity.add(b1.acceleration.times(timestep));
			b1.position.add(b1.velocity.times(timestep));
		}
		
		timeElapsed += timestep;
	}
}

File: src/physics/solvers/euler_cromer_solver.cpp
/*
 * euler_cromer_solver.cpp
 *
 *  Created on: 12 de jul de 2016
 *      Author: carlosfaruolo
 */

#include "euler_cromer_solver.hpp"

#include <vector>
using std::vector;

DEFINE_CLASS_FACTORY(EulerCromerSolver, "Semi-implicit Euler (Euler-Cromer)");

EulerCromerSolver::EulerCromerSolver(Universe2D& u)
: AbstractPhysics2DSolver(&CLASS_FACTORY, u, 0.1)
{}

void EulerCromerSolver::step()
{
	computeAccelerations();

	//velocity & position loop
	foreach(Body2D*, body, vector<Body2D*>, universe.bodies)
	{
		body->velocity += body->acceleration * timestep;
		body->position += body->velocity * timestep;
	}

	timeElapsed += timestep;
}